### PR TITLE
Fix broken Functions CLI experience for projects with incomplete GCF 2nd Gen functions.

### PR DESCRIPTION
**Pull request by taeold**
The codebase assumes that "serviceConfig" property must be present in all 2nd Gen Functions in GCF. This is an incorrect assumption - 2nd Gen GCF functions may be missing a Cloud Run service if the build never succeeded.

Also refactors `backend` module a bit to avoid calling out to Cloud Run to retrieve concurrency & cpu config - it's available on the GCF response now!

Fixes https://github.com/firebase/firebase-tools/issues/4800

### CODE REVIEW

File: src/gcp/cloudfunctionsv2.ts [hunk 1]
```diff
@@ -671,63 +669,78 @@ export function endpointFromFunction(gcfFunction: CloudFunction): backend.Endpoi
...
+        gcfFunction.serviceConfig,
+        "egressSettings",
+        "vpcConnectorEgressSettings"
+      );
+    }
+    const serviceName = gcfFunction.serviceConfig.service;
+    if (!serviceName) {
+      logger.debug(
+        "Got a v2 function without a service name." +
+          "Maybe we've migrated to using the v2 API everywhere and missed this code"
```
Comment (TheIronDev) [inline src/gcp/cloudfunctionsv2.ts:735]: Who is "we" and do we need to surface this to the end user?

Comment (TheIronDev) [reply src/gcp/cloudfunctionsv2.ts:735]: Ah, I'm seeing this was mentioned in a later line, feel free to ignore too. :+1:

**Review by TheIronDev — APPROVED** [review]

File: src/gcp/cloudfunctionsv2.ts (hunk 1 above)
Comment (taeold) [reply src/gcp/cloudfunctionsv2.ts:735]: We is us the Firebase team. I think this error might be too edge-cases too edge-casey to need to surface to the user (and hopefully not very common).

**Review by Berlioz — APPROVED** [review]
LGTM

File: src/metaprogramming.ts [hunk 2]
```diff
@@ -88,6 +88,22 @@ type DeepPickUnsafe<T, Keys extends string> = {
...
+ * type Foo = {
+ *     a?: string
+ *     b?: number
+ *     c?: object
+ * }
+ *
+ * type Bar = RequireKeys<Foo, "a" | "b">
+ * // Property "a" and "b" are now required.
+ */
+export type RequireKeys<T extends object, Keys extends keyof T> = T & {
```
Comment (inlined) [inline src/metaprogramming.ts:103]: Would you like me to make a DeepRequire like we have DeepOmit and DeepPick? Then you can say DeepRequire<Foo, "a.b" | "c">

**Pull request merged**